### PR TITLE
Add API to refresh continuous aggregate on chunk

### PR DESCRIPTION
**Pull request by k-rus**
Function refresh_continuous_aggregate, which takes a continuous
aggregate and a chunk, is added. It refreshes the continuous aggregate
on the given chunk if there are invalidations. The function can be
used in a transaction, e.g., together with following drop_chunks. This
allows users to create a user defined action to refresh and drop
chunks. Therefore, the refresh on drop is removed from drop_chunks.

### CODE REVIEW

File: tsl/test/sql/continuous_aggs_drop_chunks.sql [hunk 1]
```diff
@@ -78,37 +79,49 @@ CREATE MATERIALIZED VIEW conditions_7
...
--- is in the first chunk, which is going to be dropped, and the update will be present.
+-- is in the first chunk, which is going to be refreshed, and the update will be present.
 UPDATE conditions SET value = 4.00 WHERE time_int = 39;
 
 -- This update is in the bucket, which crosses two chunks. The first chunk is going to be 
--- dropped now and the update is outside it, and thus should not be reflected. The second 
--- chunk contains the update and will be dropped on the next call, thus it should be reflected 
--- in the continuous aggregate only after the last drop.
+-- refreshed now and the update is outside it, and thus should not be reflected. The second 
+-- chunk contains the update and will be refrehsed on the next call, thus it should be reflected
```
Comment (erimatnor) [inline tsl/test/sql/continuous_aggs_drop_chunks.sql:105]:
```suggestion
-- chunk contains the update and will be refreshed on the next call, thus it should be reflected 
```

File: tsl/src/continuous_aggs/refresh.c [hunk 2]
```diff
@@ -532,50 +543,44 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 }
 
 /*
- * Refresh all continuous aggregates on a hypertable.
+ * Refresh a continuous aggregate on a chunk.
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/refresh.c:546]:
```suggestion
 * Refresh a continuous aggregate based on the data in the given hypertable chunk.
```

File: src/chunk.c [hunk 3]
```diff
@@ -3207,32 +3218,21 @@ ts_chunk_do_drop_chunks(Hypertable *ht, int64 older_than, int64 newer_than, int3
 	{
 		int i;
 
-		/* Exclusively lock all chunks, and also refresh and invalidate the
+		/* Exclusively lock all chunks, and invalidate the
 		 * continuous aggregates in the regions covered by the chunks. Locking
 		 * prevents further modification of the dropped region during this
 		 * transaction, which allows moving the invalidation threshold without
 		 * having to worry about new invalidations while refreshing. */
```
Comment (erimatnor) [inline src/chunk.c:3225]: I think this comment needs to be updated. We are no longer moving the invalidation threshold here.

File: src/chunk.c [hunk 4]
```diff
@@ -3207,32 +3218,21 @@ ts_chunk_do_drop_chunks(Hypertable *ht, int64 older_than, int64 newer_than, int3
...
 		for (i = 0; i < num_chunks; i++)
 		{
-			int64 start = chunk_primary_dimension_start(&chunks[i]);
-			int64 end = chunk_primary_dimension_end(&chunks[i]);
+			int64 start = ts_chunk_primary_dimension_start(&chunks[i]);
+			int64 end = ts_chunk_primary_dimension_end(&chunks[i]);
 
 			LockRelationOid(chunks[i].table_id, ExclusiveLock);
```
Comment (erimatnor) [inline src/chunk.c:3231]: I am not sure we strictly need to take this lock anymore. But I guess it doesn't hurt since it will probably be taken anyway once the chunk table is dropped.

File: tsl/src/continuous_aggs/refresh.c [hunk 5]
```diff
@@ -532,50 +543,44 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
...
-	 * operation. For instance, if this is called as a result of
-	 * "refresh-on-drop", it is assumed that refresh can happen if the user is
-	 * permitted to drop data. */
+	/* Like regular materialized views, require owner to refresh. */
+	if (!pg_class_ownercheck(cagg->relid, GetUserId()))
+		aclcheck_error(ACLCHECK_NOT_OWNER,
+					   get_relkind_objtype(get_rel_relkind(cagg->relid)),
+					   get_rel_name(cagg->relid));
+
+	PreventCommandIfReadOnly(REFRESH_FUNCTION_NAME);
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/refresh.c:573]: There's a macro in `compat.h` to handle this now: `TS_PREVENT_FUNC_IF_READ_ONLY()`

File: tsl/test/sql/continuous_aggs_drop_chunks.sql [hunk 6]
```diff
@@ -78,37 +79,49 @@ CREATE MATERIALIZED VIEW conditions_7
...
+-- that the update is refreshed.
 UPDATE conditions SET value = 4.00 WHERE time_int = 2;
 
 -- This case updates values in the bucket, which affects two different partials in
--- two different chunks, which are dropped in the same call to drop_chunks.
+-- two different chunks.
 UPDATE conditions SET value = 4.00 WHERE time_int = 9;
 UPDATE conditions SET value = 4.00 WHERE time_int = 11;
 
+SELECT _timescaledb_internal.refresh_continuous_aggregate('conditions_7', show_chunks('conditions', 20));
```
Comment (erimatnor) [inline tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: I am surprised that this works given that `show_chunks` returns potentially multiple rows. Is it the case that `show_chunks` here happens to return just one row?

I would have expected this:

```suggestion
SELECT _timescaledb_internal.refresh_continuous_aggregate('conditions_7', chunk)
    FROM show_chunks('conditions', 20) chunk;
```

File: tsl/src/continuous_aggs/refresh.c [hunk 7]
```diff
@@ -364,12 +354,33 @@ continuous_agg_refresh(PG_FUNCTION_ARGS)
...
+	return cagg;
+}
 
-	cagg_ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id);
-	Assert(cagg_ht != NULL);
-	time_dim = hyperspace_get_open_dimension(cagg_ht->space, 0);
-	Assert(time_dim != NULL);
-	refresh_window.type = ts_dimension_get_partition_type(time_dim);
+static Oid
+cagg_get_partition_type(ContinuousAgg *cagg)
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/refresh.c:361]:
```suggestion
cagg_get_partition_type(const ContinuousAgg *cagg)
```

unless calls below prevent it.

**Review by erimatnor — APPROVED** [review]
LGTM, although I have a number of nits.

Also, would be good to test invalid inputs to the new function.

File: src/chunk.c [hunk 8]
```diff
@@ -195,6 +195,17 @@ chunk_formdata_fill(FormData_chunk *fd, const TupleInfo *ti)
 	if (should_free)
 		heap_freetuple(tuple);
 }
+int64
+ts_chunk_primary_dimension_start(Chunk *chunk)
```
Comment (erimatnor) [inline src/chunk.c:199]:
```suggestion
ts_chunk_primary_dimension_start(const Chunk *chunk)
```

File: tsl/src/continuous_aggs/refresh.c [hunk 9]
```diff
@@ -532,50 +543,44 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
...
-void
-continuous_agg_refresh_all(const Hypertable *const ht, const int64 start, const int64 end,
-						   const int32 chunk_id)
+Datum
+continuous_agg_refresh_chunk(PG_FUNCTION_ARGS)
 {
+	Oid cagg_relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Oid chunk_relid = PG_ARGISNULL(1) ? InvalidOid : PG_GETARG_OID(1);
+	ContinuousAgg *cagg = get_cagg_by_relid(cagg_relid);
+	Chunk *chunk = ts_chunk_get_by_relid(chunk_relid, true);
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/refresh.c:559]: I think we want to validate somehow that this chunk belongs to the hypertable that the cagg is created on and raise a useful error if not. I don't see how that happens right now.

File: tsl/src/continuous_aggs/refresh.c [hunk 10]
```diff
@@ -330,20 +330,10 @@ continuous_agg_refresh_with_window(const ContinuousAgg *cagg,
...
 }
 
-#define REFRESH_FUNCTION_NAME "refresh_continuous_aggregate()"
-/*
- * Refresh a continuous aggregate across the given window.
- */
-Datum
-continuous_agg_refresh(PG_FUNCTION_ARGS)
+static ContinuousAgg *
+get_cagg_by_relid(Oid cagg_relid)
```
Comment (pmwkaa) [inline tsl/src/continuous_aggs/refresh.c:334]: Maybe `cagg_get_by_relid` ? to be in sync with function name, as `cagg_get_partition_type`, etc

File: tsl/src/continuous_aggs/refresh.c (hunk 9 above)
Comment (pmwkaa) [reply tsl/src/continuous_aggs/refresh.c:559]: +1

File: tsl/test/sql/continuous_aggs_drop_chunks.sql (hunk 6 above)
Comment (pmwkaa) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: I also found it interesting :) I am curious if PostgreSQL has a special case for it

**Review by pmwkaa — APPROVED** [review]
Looks good overall. Is it possible to add a check to validate chunk against used cagg in the `_timescaledb_internal.refresh_continuous_aggregate` and add a test case for it?

File: src/chunk.c [hunk 11]
```diff
@@ -195,6 +195,17 @@ chunk_formdata_fill(FormData_chunk *fd, const TupleInfo *ti)
...
+{
+	return chunk->cube->slices[0]->fd.range_start;
+}
+
+int64
+ts_chunk_primary_dimension_end(Chunk *chunk)
```
Comment (mkindahl) [inline src/chunk.c]:
```suggestion
ts_chunk_primary_dimension_end(const Chunk *chunk)
```

File: src/chunk.h [hunk 12]
```diff
@@ -162,14 +162,14 @@ extern TSDLLEXPORT List *ts_chunk_get_chunk_ids_by_hypertable_id(int32 hypertabl
 extern TSDLLEXPORT List *ts_chunk_get_data_node_name_list(const Chunk *chunk);
 extern List *ts_chunk_data_nodes_copy(const Chunk *chunk);
 
+extern TSDLLEXPORT int64 ts_chunk_primary_dimension_start(Chunk *chunk);
+
+extern TSDLLEXPORT int64 ts_chunk_primary_dimension_end(Chunk *chunk);
```
Comment (mkindahl) [inline src/chunk.h]:
```suggestion
extern TSDLLEXPORT int64 ts_chunk_primary_dimension_start(const Chunk *chunk);

extern TSDLLEXPORT int64 ts_chunk_primary_dimension_end(const Chunk *chunk);
```

File: tsl/src/continuous_aggs/refresh.c [hunk 13]
```diff
@@ -364,12 +354,33 @@ continuous_agg_refresh(PG_FUNCTION_ARGS)
...
+{
+	Hypertable *cagg_ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id);
+	Dimension *time_dim = hyperspace_get_open_dimension(cagg_ht->space, 0);
+
+	return ts_dimension_get_partition_type(time_dim);
+}
+
+#define REFRESH_FUNCTION_NAME "refresh_continuous_aggregate()"
+/*
+ * Refresh a continuous aggregate across the given window.
```
Comment (mkindahl) [inline tsl/src/continuous_aggs/refresh.c:371]:
```suggestion
 * Refresh a continuous aggregate in the given window.
```

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 14]
```diff
@@ -186,15 +187,23 @@ SELECT * FROM see_cagg;
...
 -- This is the simplest case, when the updated bucket is inside a chunk, so it is expected
--- that the update is always reflected after the refresh on drop.
+-- that the update is refreshed.
 UPDATE conditions SET value = 4.00 WHERE time_int = 2;
 -- This case updates values in the bucket, which affects two different partials in
--- two different chunks, which are dropped in the same call to drop_chunks.
+-- two different chunks.
 UPDATE conditions SET value = 4.00 WHERE time_int = 9;
 UPDATE conditions SET value = 4.00 WHERE time_int = 11;
+SELECT _timescaledb_internal.refresh_continuous_aggregate('conditions_7', show_chunks('conditions', 20));
```
Comment (mkindahl) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:200]:
```suggestion
CALL _timescaledb_internal.refresh_continuous_aggregate('conditions_7', show_chunks('conditions', 20));
```

File: sql/ddl_internal.sql [hunk 15]
```diff
@@ -20,3 +20,8 @@ AS '@MODULE_PATHNAME@', 'ts_data_node_block_new_chunks' LANGUAGE C VOLATILE;
 -- distributed hypertables
 CREATE OR REPLACE FUNCTION _timescaledb_internal.allow_new_chunks(data_node_name NAME, hypertable REGCLASS = NULL) RETURNS INTEGER
 AS '@MODULE_PATHNAME@', 'ts_data_node_allow_new_chunks' LANGUAGE C VOLATILE;
+
+CREATE OR REPLACE FUNCTION _timescaledb_internal.refresh_continuous_aggregate(
```
Comment (mkindahl) [inline sql/ddl_internal.sql:24]: Since the public version is a procedure, this one should probably also be a procedure. The return value is not useful anyway.
```suggestion
CREATE OR REPLACE PROCEDURE _timescaledb_internal.refresh_continuous_aggregate(
```

File: tsl/test/sql/continuous_aggs_drop_chunks.sql (hunk 6 above)
Comment (mkindahl) [inline tsl/test/sql/continuous_aggs_drop_chunks.sql:95]:
```suggestion
CALL _timescaledb_internal.refresh_continuous_aggregate('conditions_7', show_chunks('conditions', 20));
```

Comment (k-rus) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: @mkindahl Will it work with `show_chunks`, which returns set of chunks? How will the procedure work with selecting all continuous aggregates of the hypertable?
Combining the refresh function with set of chunks and set of caggs is the reason to use the function and not the procedure. May be I miss something and it can be also simple with procedure?

Comment (mkindahl) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: AFAICT, there is no real difference between functions and procedures internally, so I would be surprised if it did not work, but the only way to know for sure is to test it.

Comment (k-rus) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: The difference is in the usage. Procedures cannot be used in select statements, thus the query planner will not resolve calling show_chunks, since it returns a set. Your suggestion will require to use cursor, I guess.

Comment (mkindahl) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: I tested it, and you're right. The set-returning function cannot be used with the procedure, but a function works.

Comment (mkindahl) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]:
```
mats=# CREATE TABLE department(id int primary key, name text);
CREATE TABLE
mats=# CREATE TABLE employee(id int primary key, name text, salary int, departmentid int references department);
CREATE TABLE
mats=# INSERT INTO department VALUES (1, 'Management');
INSERT 0 1
mats=# INSERT INTO department VALUES (2, 'IT');
INSERT 0 1
mats=# INSERT INTO employee VALUES (1, 'John Smith', 30000, 1);
INSERT 0 1
mats=# INSERT INTO employee VALUES (2, 'Jane Doe', 50000, 1);
INSERT 0 1
mats=# INSERT INTO employee VALUES (3, 'Jack Jackson', 60000, 2);
INSERT 0 1
mats=# CREATE TABLE log(id int, message text);
CREATE TABLE
mats=# CREATE PROCEDURE LogName(int) AS $$
mats$#    INSERT INTO log VALUES ($1, format('%s was here', EmployeeName($1)))
mats$# $$ LANGUAGE sql;
CREATE PROCEDURE
mats=# CREATE FUNCTION EmployeeName(int) RETURNS text AS 'SELECT name FROM employee WHERE id = $1' LANGUAGE sql;
CREATE FUNCTION
mats=# CREATE FUNCTION GetEmployees() RETURNS SETOF int AS 'select id from employee;' language sql;
CREATE FUNCTION
mats=# CALL LogName(GetEmployees());
ERROR:  set-returning functions are not allowed in CALL arguments
LINE 1: CALL LogName(GetEmployees());
                     ^
mats=# DROP PROCEDURE LogName;
DROP PROCEDURE
mats=# CREATE FUNCTION LogName(int) RETURNS VOID AS $$
mats$#    INSERT INTO log VALUES ($1, format('%s was here', EmployeeName($1)))
mats$# $$ LANGUAGE sql;
CREATE FUNCTION
mats=# SELECT LogName(GetEmployees());
 logname 
---------
 [NULL]
 [NULL]
 [NULL]
(3 rows)

mats=# SELECT * FROM log;
 id |        message        
----+-----------------------
  1 | John Smith was here
  2 | Jane Doe was here
  3 | Jack Jackson was here
(3 rows)
```

File: sql/ddl_internal.sql (hunk 15 above)
Comment (k-rus) [reply sql/ddl_internal.sql:24]: But it cannot be used in queries and, since it takes only single continuous aggregate and single chunk, it will be cumbersome to use. I suggest to improve the name of the function instead, but after this PR, before moving the function to public space.

Comment (mkindahl) [reply sql/ddl_internal.sql:24]: Makes sense. With the info below I think we should keep this as a function.

File: tsl/test/sql/continuous_aggs_drop_chunks.sql (hunk 6 above)
Comment (k-rus) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: It is working on a set. PG query planner knows how to plan such query correctly. We have already plenty examples including docs with other single node functions.

Comment (k-rus) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:95]: See in this PR, which means that 2 chunks had invalidations:
```
    SELECT _timescaledb_internal.refresh_continuous_aggregate('conditions_7', show_chunks('conditions', 40));
   refresh_continuous_aggregate 
  ------------------------------
 
 
  (2 rows)
```

File: sql/ddl_internal.sql (hunk 15 above)
Comment (gayyappan) [reply sql/ddl_internal.sql:24]: @k-rus I am confused by this comment. If this is a function, it can be used in queries. 
Could we modify the name now instead of deferring? We now have 1 function and a procedure with the same name and it is not clear how they differ in functionality. We would certainly document this with an example in the docs. Having a better name would really help.

Comment (k-rus) [reply sql/ddl_internal.sql:24]: @gayyappan There is no better proposed name. I made another name in the proposal, which I shared on Slack in database channel last week. Mine name there is not good either.

Since the function is internal, we don't need to deliver good name immediately.

Comment (gayyappan) [reply sql/ddl_internal.sql:24]: My suggestion for a name is refresh_continuous_aggregate_for_raw_chunk_boundary. It is a mouthful but I think the semantics is clear. Since it is internal , we could shorten this to refresh_cagg_for_raw_chunk.

**Review by mkindahl — APPROVED** [review]

File: sql/ddl_internal.sql (hunk 15 above)
Comment (erimatnor) [reply sql/ddl_internal.sql:24]: FWIW, IMO this function does the exact same thing as the other `refresh_continuous_aggregate`: namely it refreshes the cagg in the region specified. In one case the region is specified by an explicit window (start, end), in the second case the region is implicit by the chunk (still a start, end).

To only other difference is that the function in this PR can be used in a transaction, while the other procedure cannot (currently). But there is no reason why we cannot make the window-version transactional if it is executed within a transaction block.
